modules/FvwmCpp/FvwmCpp.1
.\" t
.\" @(#)FvwmCpp.1	1/28/94
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.5i
..
.de EE
.fi
.in -.5i
.if n .sp 1
.if t .sp .5
..
.TH FvwmCpp 1 "7 May 1999"
.UC
.SH NAME
FvwmCpp \- the FVWM  Cpp pre-processor
.SH SYNOPSIS
FvwmCpp is spawned by fvwm, so no command line invocation will work.

.SH DESCRIPTION
When called, this module will attempt to have /usr/lib/cpp pre-process
the file specified in its invocation, and then have fvwm read the
resulting file.

.SH INVOCATION
FvwmCpp can be invoked by inserting the line 'FvwmCpp' in the .fvwm2rc
file. It can also be called from a menu or mouse binding.  If the user
wants his entire .fvwm2rc file pre-processed with FvwmCpp, then fvwm
should be invoked as:

.EX
fvwm2 -cmd "FvwmCpp .fvwm2rc"
.EE

Some options can be specified on the command line:
.TP
-cppopt \fIoption\fP
Lets you pass an option to the cpp program.  Not really needed as any unknown
options will be passed on automatically.

.TP
-cppprog \fIname\fP
Instead of invoking "/usr/lib/cpp", fvwm will invoke \fIname\fP.

.TP
-outfile \fIfilename\fP
Instead of creating a random unique name for the temporary file for
the preprocessed rc file, this option will let you specify the name of
the temporary file it will create.  Please not that FvwmCpp will attempt
to remove this file before writing to it, so don't point it at anything
important even if has read-only protection.

.IP -debug
Causes the temporary file create by Cpp to
be retained. This file is usually called "/tmp/fvwmrcXXXXXX"

.IP -lock
If you want to use this option you need to start FvwmCpp with
ModuleSynchronous. This option causes fvwm to wait that the pre-process
finish and that FvwmCpp asks fvwm to Read the pre-processed file before
continuing. This may be useful at startup if you use a session manager
as Gnome. Also, this is useful if you want to process and run a Form in
a fvwm function.

.IP -noread
Causes the pre-processed file to be not read by fvwm. Useful to
pre-process a FvwmScript script with FvwmCpp.

.SH CONFIGURATION OPTIONS
FvwmCpp defines some values for use in the pre-processor file:

.IP TWM_TYPE
Always set to "fvwm".
.IP SERVERHOST
The name of the machine running the X Server.
.IP CLIENTHOST
The name of the machine running fvwm.
.IP HOSTNAME
The host name of the machine running fvwm. Generally the same as CLIENTHOST.
.IP OSTYPE
The operating system for CLIENTHOST.
.IP USER
The name of the person running fvwm.
.IP HOME
The home directory of the person running fvwm.
.IP VERSION
The X11 version.
.IP REVISION
The X11 revision number.
.IP VENDOR
The X server vendor.
.IP RELEASE
The X server release number.
.IP SCREEN
The screen number.
.IP WIDTH
The screen width in pixels.
.IP HEIGHT
The screen height in pixels.
.IP X_RESOLUTION
Some distance/pixel measurement for the horizontal direction, I think.
.IP Y_RESOLUTION
Some distance/pixel measurement for the vertical direction, I think.
.IP PLANES
Number of color planes for the X server display
.IP BITS_PER_RGB
Number of bits in each rgb triplet.
.IP CLASS
The X11 default visual class, e.g. PseudoColor.
.IP COLOR
Yes or No, Yes if the default visual class is neither StaticGrey or GreyScale.
.IP FVWM_CLASS
The visual class that fvwm is using, e.g. TrueColor.
.IP FVWM_COLOR
Yes or No, Yes if the FVWM_CLASS is neither StaticGrey or GreyScale.
.IP FVWM_VERSION
The fvwm version number, ie 2.0
.IP OPTIONS
Some combination of SHAPE, XPM, NO_SAVEUNDERS, and Cpp, as defined in
configure.h at compile time.
.IP FVWM_MODULEDIR
The directory where fvwm looks for .fvwmrc and modules by default, as
determined at compile time.
.IP FVWM_USERDIR
The value of $FVWM_USERDIR.
.IP SESSION_MANAGER
The value of $SESSION_MANAGER. Undefined if this variable is not set.

.SH EXAMPLE PROLOG

.EX
#define TWM_TYPE fvwm
#define SERVERHOST spx20
#define CLIENTHOST grumpy
#define HOSTNAME grumpy
#define OSTYPE SunOS
#define USER nation
#define HOME /local/homes/dsp/nation
#define VERSION 11
#define REVISION 0
#define VENDOR HDS human designed systems, inc. (2.1.2-D)
#define RELEASE 4
#define SCREEN 0
#define WIDTH 1280
#define HEIGHT 1024
#define X_RESOLUTION 3938
#define Y_RESOLUTION 3938
#define PLANES 8
#define BITS_PER_RGB 8
#define CLASS PseudoColor
#define COLOR Yes
#define FVWM_VERSION 2.0 pl 1
#define OPTIONS SHAPE XPM Cpp
#define FVWM_MODULEDIR /local/homes/dsp/nation/modules
#define FVWM_USERDIR /local/homes/dsp/nation/.fvwm2rc.d
#define SESSION_MANAGER local/grumpy:/tmp/.ICE-unix/440,tcp/spx20:1025
.EE

.SH BUGS
Module configurations do not become active until fvwm has restarted
if you use FvwmCpp on startup. FvwmCpp creates a temporary file
and passes this to fvwm, so you would have to edit this file too.
There are some problems with comments in your .fvwm2rc file.
The comment sign # is misinterpreted by the preprocessor.
This has usually no impact on functionality but generates
annoying warning messages.
The sequence /* is interpreted as the start of a C comment what
is probably not what you want in a filename. You might want to try
/?* (for filenames only) or /\\* or "/*" instead. Depending on
your preprocessor you may have the same problem with "//".
Macros are not replaced within single (') or double quotes (\"). Use
back quotes (`) to circumvent this. Fvwm accepts back quotes for
quoting and at least FvwmButtons does too.
The preprocessor may place a space after a macro substitution, so
with
.EX
#define MYCOMMAND ls
"Exec "MYCOMMAND" -l"
.EE
you might get
.EX
"Exec "ls " -l" (two words)
.EE
and not
.EX
"Exec "ls" -l" (one word).
.EE

If you use gcc you can use this invocation to turn off '//'
comments:

.EX
FvwmCpp -Cppprog '/your/path/to/gcc -C -E -' <filename>
.EE

.SH AUTHOR
FvwmCpp is the result of a random bit mutation on a hard disk,
presumably a result of a  cosmic-ray or some such thing.


NEWS
Note, the changes for the last STABLE release start with release
2.6.0.

-------------------------------------------------------------------
Changes in stable release 2.6.8 (UNRELEASED)

* Bug fixes:

  - Various DESTDIR fixes (especially around the default-config
    Makefile)
  - fvwm-perllib pod2man fixes.
  - FvwmIconMan no longer triggers a warning about bad size hints
    in the fvwm core.
  - VMware windows do not disapper when switching between
    fullscreen and normal state.
  - Fix (de)installation with the configure options
    --program-prefix, --program-suffix and
    --program-transform-name.

-------------------------------------------------------------------
Changes in stable release 2.6.7 (06-Mar-2016)

* New fvwm features:

  - A new default configuration which is available when fvwm
    doesn't detect a configuration file to load.
  - A new conitional command "Desk n" can restrict matching
    windows to a specific desk.
  - A new conditional command "Screen n" to restrict matching
    windows on a given Xinerama screen.
  - New expansion variable "w.screen" to ascertain the Xineram
    screen number a window is on.
  - New command "InfoStoreClear" to remove all items in the
    InfoStore.

* Removed features:

  - The old and unmaintained debian/and rpm/ directories have
    been remmoved; use the maintainers' copies where available.
  - VMS support has been removed.
  - GTK1.x support has been removed.
  - GNOME-specific window hints (pre-EWMH) have been removed.
  - Some fvwm modules have been removed:
      - FvwmDragWell   (no replacement)
      - FvwmGTK        (no replacement)
      - FvwmSave       (no replacement)
      - FvwmSaveDesk   (no replacement)
      - FvwmScroll     (no replacement)
      - FvwmTabs       (no replacement, never worked anyway)
      - FvwmTaskBar    (use FvwmButtons)
      - FvwmTheme      (in core of fvwm as colorsets)
      - FvwmWharf      (use FvwmButtons)
      - FvwmWinList    (use WindowList command)
      - FvwmWindowMenu (use WindowList command)
      - FvwmIconBox    (use the IconBox style instead)

* New module features:

  - FvwmButtons learned a new option "Colorset" to its
    ChangeButton command.

* fvwm-menu-desktop updated:

  - Renamed default menu to XDGMenu and changed the name
    of the FvwmForm to FvwmForm-XDGMenu-Config to not conflict
    with someone already using FvwmMenu.
  - fvwm-menu-desktop will now load defaults from the
    FvwmForm-XDGMenu-Config data file.
  - Improved dynamic menus by regenerating them on-the-fly.
  - Added new options: --regen-cmd, --dynamic, and more.

* Bug fixes:

  - A bug introduced in 2.6.6 could cause applications with
    negative coordinates to be placed at strange positions.  This
    affected for example acroread when switching to fullscreen
    mode.  This has been fixed.
  - FvwmButtons "Silent" option for dynamic ChangeButton commands
    no longer loops infinitely.

Changes in stable release 2.6.6 (15-Mar-2016)

* New fvwm features:

  - Support for Russian from Ivan Gayevskiy.
  - EnvMatch supports infostore variables.
  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.
  - Windows shaded to a corner (NW, NE, SW, SE) are reduced to a
    small square.
  - New option "!raise" to the WarpToWindow command.
  - The new extended variables $[wa.x], $[wa.y], $[wa.width],
    $[wa.height] can be used to get the geometry of the EWMH
    working area, and $[dwa.x], $[dwa.y], $[dwa.width],
    $[dwa.height] can be used to get the geometry of the EWMH
    dynamic working area.
  - The Resize commands accept "wa" or "da" as a suffix of the
    width or height value.  If present, the value is a percentage
    of the width or height of the EWMH working area or the EWMH
    dynamic working area.
  - Fvwm is much more resilient against applications that flood the
    window manager with repeated events.

* New module features:

  - FvwmForm supports separator lines.
  - New FvwmIconMan options:
      IconAndSelectButton
      IconAndSelectColorset

* Bug fixes:

  - Provide a wrapper for the deprecation of XKeycodeToKeysym and
    use XkbKeycodeToKeysym() where appropriate.
  - fvwm-menu-desktop is re-written and provides better support of
    the XDG menu specification.
  - Fix fvwm-menu-desktop keyError bug.  Use "others" if no desktop
    environment found.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Globally active windows cannot take the focus if the style
    forbids programs to take focus themselves (style
    !FPFocusByProgram).
  - Windows no longer jump from one position to the other which
    could happen in some cases with SnapAttraction.  Windows now
    snap to the closest window (or screen edge).
  - Removing bindings had several strange side effects that are
    fixed now (removing too many bindings; old bindings showing up
    again after another is removed; possibly other effects).
  - Windows sometimes did not get expose events (i.e. did not
    redraw properly) if they were uncovered by moving a window
    above them.  This has been fixed.
  - FvwmConsole now causes much less network traffic.
  - Suppress bogus events sent to the modules when a window is
    resized with the mouse.
  - Properly handle the has_ref_window_moved flag for ResizeMove and
    ResizeMoveMaximize.
  - Removed some unnecessary redraws in FvwmPager.
  - The option "NoDeskLimitY" option of the GotoPage command did
    not work.
  - Negative coordinates in the "rectangle" option to the Menu
    commend did not work correctly.  This has been fixed.
  - Removes a slight graphics problem whith the ResizeMaximize
    command being invoked from a window button menu.
  - When an attempt to reparent a client window (i.e. decorate
    it) fails, fvwm no longer throws away all events but only the
    events for that window.

modules/FvwmCpp/FvwmCpp.c
/* -*-c-*- */
/* This module, and the entire FvwmCpp program, and the concept for
 * interfacing this module to the Window Manager, are all original work
 * by Robert Nation
 */

/* This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "config.h"

#include <stdio.h>
#include <signal.h>
#include <fcntl.h>
#include <sys/wait.h>
#include "libs/ftime.h"
#include <unistd.h>
#include <ctype.h>
#include <pwd.h>
#include <sys/param.h>
#include <sys/types.h>
#include <netdb.h>

#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xproto.h>
#include <X11/Xatom.h>
#include <X11/Intrinsic.h>

#include "libs/Module.h"
#include "libs/fvwm_sys_stat.h"

#include "FvwmCpp.h"
#include "libs/fvwmlib.h"
#include "libs/FShape.h"
#include "libs/PictureBase.h"
#include "libs/FSMlib.h"
#include "libs/Strings.h"
#include "libs/System.h"
#include <X11/StringDefs.h>
#include <X11/Shell.h>
#define Resolution(pixels, mm) ((((pixels) * 2000 / (mm)) + 1) / 2)

char *MyName;
int fd[2];

long Vx, Vy;
static char *MkDef(char *name, char *def);
static char *MkNum(char *name,int def);
static char *cpp_defs(Display *display, const char *host, char *m4_options, char *config_file);
#define MAXHOSTNAME 255
#define EXTRA 20

char *cpp_prog = FVWM_CPP;          /* Name of the cpp program */

char cpp_options[BUFSIZ];
char cpp_outfile[BUFSIZ]="";


/*
 *
 *  Procedure:
 *      main - start of module
 *
 */
int main(int argc, char **argv)
{
  Display *dpy;                 /* which display are we talking to */
  char *temp, *s;
  char *display_name = NULL;
  char *filename = NULL;
  char *tmp_file;
  int i;
  int debug = 0;
  int lock = 0;
  int noread = 0;
  char *user_dir;

  /* Figure out the working directory and go to it */
  user_dir = getenv("FVWM_USERDIR");
  if (user_dir != NULL)
  {
    if (chdir(user_dir) < 0)
      fprintf(stderr, "%s: <<Warning>> chdir to %s failed in cpp_defs",
	      MyName, user_dir);
  }

  sprintf(cpp_options, "-I'%s' ", FVWM_DATADIR);

  /* Record the program name for error messages */
  temp = argv[0];

  s=strrchr(argv[0], '/');
  if (s != NULL)
    temp = s + 1;

  MyName = safemalloc(strlen(temp)+2);
  strcpy(MyName,"*");
  strcat(MyName, temp);

  if(argc < 6)
    {
      fprintf(stderr,"%s Version %s should only be executed by fvwm!\n",MyName,
	      VERSION);
      exit(1);
    }

  /* We should exit if our fvwm pipes die */
  signal (SIGPIPE, DeadPipe);

  fd[0] = atoi(argv[1]);
  fd[1] = atoi(argv[2]);

  for(i=6;i<argc;i++)
    {
      /* leaving this in just in case-- any option starting with '-'
	 will get passed on to cpp anyway */
      if(strcasecmp(argv[i],"-cppopt") == 0)
	{
	  strcat(cpp_options, argv[++i]);
	  strcat(cpp_options, " ");
	}
      else if (strcasecmp(argv[i], "-cppprog") == 0)
	{
	  cpp_prog = argv[++i];
	}
      else if(strcasecmp(argv[i], "-outfile") == 0)
      {
	strcpy(cpp_outfile,argv[++i]);
      }
      else if(strcasecmp(argv[i], "-debug") == 0)
	{
	  debug = 1;
	}
      else if(strcasecmp(argv[i], "-lock") == 0)
	{
	  lock = 1;
	}
      else if(strcasecmp(argv[i], "-noread") == 0)
	{
	  noread = 1;
	}
      else if (strncasecmp(argv[i],"-",1) == 0)
      {
	/* pass on any other arguments starting with '-' to cpp */
	strcat(cpp_options, argv[i]);
	strcat(cpp_options, " ");
      }
      else
	filename = argv[i];
    }

  /* If we don't have a cpp, stop right now.  */
  if (!cpp_prog || cpp_prog[0] == '\0') {
    fprintf(stderr, "%s: no C preprocessor program specified\n", MyName);
    exit(1);
  }

  if (!filename)
  {
    fprintf(stderr, "%s: no file specified.\n", MyName);
    exit(1);
  }
  for(i=0;i<strlen(filename);i++)
    if((filename[i] == '\n')||(filename[i] == '\r'))
      {
	filename[i] = 0;
      }

  if (!(dpy = XOpenDisplay(display_name)))
    {
      fprintf(stderr,"%s: can't open display %s",
	      MyName, XDisplayName(display_name));
      exit (1);
    }

  /* set up G */
  PictureInitCMap(dpy);

  /* tell fvwm we're running if -lock is not used */
  if (!lock)
    SendFinishedStartupNotification(fd);

  tmp_file = cpp_defs(dpy, display_name,cpp_options, filename);

  if (!noread)
  {
    char *read_string = CatString3("Read '", tmp_file, "'");
    SendText(fd, read_string, 0);
  }

  /* tell fvwm to continue if -lock is used */
  if (lock)
    SendFinishedStartupNotification(fd);

  /* For a debugging version, we may wish to omit this part. */
  /* I'll let some cpp advocates clean this up */
  if (!debug)
  {
    char *delete_string;
    char *delete_file = tmp_file;
    if (tmp_file[0] != '/' && user_dir != NULL)
    {
      delete_file = safestrdup(CatString3(user_dir, "/", tmp_file));
    }
    delete_string = CatString3("Exec exec /bin/rm '", delete_file, "'");
    SendText(fd, delete_string, 0);
  }

  return 0;
}



static char *cpp_defs(Display *display, const char *host, char *cpp_options, char *config_file)
{
	Screen *screen;
	Visual *visual;
	char client[MAXHOSTNAME], server[MAXHOSTNAME], *colon;
	char ostype[BUFSIZ];
	char options[BUFSIZ];
	static char tmp_name[BUFSIZ];
	struct hostent *hostname;
	char *vc;                     /* Visual Class */
	FILE *tmpf;
	struct passwd *pwent;
	int fd;
	int Mscreen;

	/* Generate a temporary filename.  Honor the TMPDIR environment variable,
	   if set. Hope nobody deletes this file! */

	if (strlen(cpp_outfile) == 0)
	{
		if ((vc=getenv("TMPDIR")))
		{
			strcpy(tmp_name, vc);
		}
		else
		{
			strcpy(tmp_name, "/tmp");
		}
		strcat(tmp_name, "/fvwmrcXXXXXX");
		fd = fvwm_mkstemp(tmp_name);
		if (fd == -1)
		{
			fprintf(
				stderr,
				"[FvwmCpp][cpp_def] fvwm_mkstemp failed %s\n",
				tmp_name);
			exit(0377);
		}
	}
	else
	{
		strcpy(tmp_name,cpp_outfile);
		/*
		 * check to make sure it doesn't exist already, to prevent
		 * security hole
		 */
		/* first try to unlink it */
		unlink(tmp_name);
		fd = open(
			tmp_name, O_WRONLY|O_EXCL|O_CREAT,
			FVWM_S_IRUSR | FVWM_S_IWUSR);
		if (fd < 0)
		{
			fprintf(
				stderr,
				"[FvwmCpp][cpp_defs] error opening file %s\n",
				tmp_name);
			exit(0377);
		}
	}
	close(fd);

	/*
	 * Create the appropriate command line to run cpp, and
	 * open a pipe to the command.
	 */

	sprintf(options, "%s %s >%s\n", cpp_prog, cpp_options, tmp_name);
	tmpf = popen(options, "w");
	if (tmpf == NULL)
	{
		fprintf(
			stderr,
			"[FvwmCpp][cpp_defs] Cannot open pipe to cpp\n");
		exit(0377);
	}

	gethostname(client,MAXHOSTNAME);

	getostype  (ostype, sizeof ostype);

	hostname = gethostbyname(client);
	strcpy(server, XDisplayName(host));
	colon = strchr(server, ':');
	if (colon != NULL) *colon = '\0';
	if ((server[0] == '\0') || (!strcmp(server, "unix")))
		strcpy(server, client); /* must be connected to :0 or unix:0 */

	/* TWM_TYPE is fvwm, for completeness */

	fputs(MkDef("TWM_TYPE", "fvwm"), tmpf);

	/* The machine running the X server */
	fputs(MkDef("SERVERHOST", server), tmpf);
	/* The machine running the window manager process */
	fputs(MkDef("CLIENTHOST", client), tmpf);
	if (hostname)
		fputs(MkDef("HOSTNAME", (char *)hostname->h_name), tmpf);
	else
		fputs(MkDef("HOSTNAME", (char *)client), tmpf);

	fputs(MkDef("OSTYPE", ostype), tmpf);

	pwent=getpwuid(geteuid());
	fputs(MkDef("USER", pwent->pw_name), tmpf);

	fputs(MkDef("HOME", getenv("HOME")), tmpf);
	fputs(MkNum("VERSION", ProtocolVersion(display)), tmpf);
	fputs(MkNum("REVISION", ProtocolRevision(display)), tmpf);
	fputs(MkDef("VENDOR", ServerVendor(display)), tmpf);
	fputs(MkNum("RELEASE", VendorRelease(display)), tmpf);

	Mscreen= DefaultScreen(display);
	fputs(MkNum("SCREEN", Mscreen), tmpf);

	fputs(MkNum("WIDTH", DisplayWidth(display,Mscreen)), tmpf);
	fputs(MkNum("HEIGHT", DisplayHeight(display,Mscreen)), tmpf);

	screen = ScreenOfDisplay(display, Mscreen);
	fputs(MkNum("X_RESOLUTION",Resolution(screen->width,screen->mwidth)),
	      tmpf);
	fputs(MkNum("Y_RESOLUTION",Resolution(screen->height,screen->mheight)),
	      tmpf);
	fputs(MkNum("PLANES",DisplayPlanes(display, Mscreen)), tmpf);

	visual = DefaultVisualOfScreen(screen);
	fputs(MkNum("BITS_PER_RGB", visual->bits_per_rgb), tmpf);

	switch(visual->class)
	{
	case(StaticGray):
		vc = "StaticGray";
		break;
	case(GrayScale):
		vc = "GrayScale";
		break;
	case(StaticColor):
		vc = "StaticColor";
		break;
	case(PseudoColor):
		vc = "PseudoColor";
		break;
	case(TrueColor):
		vc = "TrueColor";
		break;
	case(DirectColor):
		vc = "DirectColor";
		break;
	default:
		vc = "NonStandard";
		break;
	}
	fputs(MkDef("CLASS", vc), tmpf);

	switch(Pvisual->class)
	{
	case(StaticGray):
		vc = "StaticGray";
		break;
	case(GrayScale):
		vc = "GrayScale";
		break;
	case(StaticColor):
		vc = "StaticColor";
		break;
	case(PseudoColor):
		vc = "PseudoColor";
		break;
	case(TrueColor):
		vc = "TrueColor";
		break;
	case(DirectColor):
		vc = "DirectColor";
		break;
	default:
		vc = "NonStandard";
		break;
	}
	fputs(MkDef("FVWM_CLASS", vc), tmpf);

	if (visual->class != StaticGray && visual->class != GrayScale)
		fputs(MkDef("COLOR", "Yes"), tmpf);
	else
		fputs(MkDef("COLOR", "No"), tmpf);

	if (Pvisual->class != StaticGray && Pvisual->class != GrayScale)
		fputs(MkDef("FVWM_COLOR", "Yes"), tmpf);
	else
		fputs(MkDef("FVWM_COLOR", "No"), tmpf);

	fputs(MkDef("FVWM_VERSION", VERSION), tmpf);

	/* Add options together */
	*options = '\0';
	if (FHaveShapeExtension)
		strcat(options, "SHAPE ");

	if (XpmSupport)
		strcat(options, "XPM ");

	strcat(options, "Cpp ");

	fputs(MkDef("OPTIONS", options), tmpf);

	fputs(MkDef("FVWM_MODULEDIR", FVWM_MODULEDIR), tmpf);
	fputs(MkDef("FVWM_DATADIR", FVWM_DATADIR), tmpf);

	if ((vc = getenv("FVWM_USERDIR")))
		fputs(MkDef("FVWM_USERDIR", vc), tmpf);

	if (SessionSupport && (vc = getenv("SESSION_MANAGER")))
		fputs(MkDef("SESSION_MANAGER", vc), tmpf);


	/*
	 * At this point, we've sent the definitions to cpp.  Just include
	 * the fvwmrc file now.
	 */

	fprintf(tmpf, "#include \"%s\"\n", config_file);

	pclose(tmpf);
	return(tmp_name);
}





/*
 *
 *  Procedure:
 *      SIGPIPE handler - SIGPIPE means fvwm is dying
 *
 */
RETSIGTYPE DeadPipe(int nonsense)
{
  exit(0);
  SIGNAL_RETURN;
}

static char *MkDef(char *name, char *def)
{
  char *cp = NULL;
  int n;

  /* Get space to hold everything, if needed */

  n = EXTRA + strlen(name) + strlen(def);
  cp = safemalloc(n);

  sprintf(cp, "#define %s %s\n",name,def);

  return(cp);
}

static char *MkNum(char *name,int def)
{
  char num[20];

  sprintf(num, "%d", def);

  return(MkDef(name, num));
}
